cub/test/catch2_test_device_histogram.cu

```diff
@@ -0,0 +1,662 @@
+    std::copy(h_levels.begin(), h_levels.end(), d_levels.begin());
+    CAPTURE(d_levels);
+
+    // Compute reference result
+    const auto sample_to_bin_index = [&](int channel, SampleT sample) {
+      const auto* l = h_levels[channel].data();
+      const auto n  = h_levels[channel].size();
+      const int bin = static_cast<int>(
+        std::distance(l, std::upper_bound(l, l + n, static_cast<LevelT>(sample))) - 1); // TODO(bgruber):
+                                                                                        // why not
+                                                                                        // lower_bound?
+      return bin < 0 ? n : bin;
+    };
+    auto h_histogram = compute_reference_result<Channels, CounterT>(
+      h_samples, sample_to_bin_index, num_levels, width, height, row_pitch);
+
+    // Compute result and verify
+    _CCCL_IF_CONSTEXPR (ActiveChannels == 1 && Channels == 1)
+    {
+      histogram_range(
+        cast_if_half_pointer(thrust::raw_pointer_cast(d_samples.data())),
+        to_ptr_array(d_histogram)[0],
+        d_levels[0].size(),
+        cast_if_half_pointer(to_ptr_array(d_levels).data())[0],
+        width,
+        height,
+        row_pitch);
+    }
+    else
+    {
+      multi_histogram_range<Channels, ActiveChannels>(
+        cast_if_half_pointer(thrust::raw_pointer_cast(d_samples.data())),
+        to_ptr_array(d_histogram).data(),
+        to_size_array(d_levels).data(),
+        cast_if_half_pointer(to_ptr_array(d_levels).data()),
+        width,
+        height,
+        row_pitch);
+    }
+    for (std::size_t c = 0; c < ActiveChannels; ++c)
+    {
+      CHECK(h_histogram[c] == d_histogram[c]);
+    }
+  }
+}
+
+using types =
+  c2h::type_list<std::int8_t,
+                 std::uint8_t,
+                 std::int16_t,
+                 std::uint16_t,
+                 std::int32_t,
+                 std::uint32_t,
+                 std::int64_t,
+                 std::uint64_t,
+#if TEST_HALF_T
+                 half_t,
+#endif
+                 float,
+                 double>;
+
+CUB_TEST("DeviceHistogram::Histogram* basic use", "[histogram][device]", types)
+{
+  using sample_t = c2h::get<0, TestType>;
+  using level_t =
+    typename std::conditional<cub::NumericTraits<sample_t>::CATEGORY == cub::FLOATING_POINT, sample_t, int>::type;
+  // Max for int8/uint8 is 2^8, for half_t is 2^10. Beyond, we would need a different level generation
+  const auto max_level       = level_t{sizeof(sample_t) == 1 ? 256 : 1024};
+  const auto max_level_count = (sizeof(sample_t) == 1 ? 256 : 1024) + 1;
+  test_even_and_range<sample_t, 4, 3, int>(max_level, max_level_count, 1920, 1080);
+}
+
+// TODO(bgruber): float produces INFs in the HistogramRange test setup AND the HistogramEven implementation
+// This test covers int32 and int64 arithmetic for bin computation
+CUB_TEST("DeviceHistogram::Histogram* large levels", "[histogram][device]", metal::remove<types, float>)
```

Comment:
This test fails for `float`, both computing the reference result and the CUB histogram implementation. Should I investiage this as part of this PR?

Reply:
I would say add an issue and follow up on this PR

Reply:
Alright, here is the issue: #1793